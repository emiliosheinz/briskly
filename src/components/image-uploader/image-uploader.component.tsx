import React, { useState } from 'react'

import { DocumentPlusIcon } from '@heroicons/react/24/outline'

import { useFilePreview } from '~/hooks/use-file-preview'

import { Image } from '../image/image.component'
import type { ImageUploaderProps } from './image-uploader.types'

export const ImageUploader = React.forwardRef<
  HTMLInputElement,
  ImageUploaderProps
>(function ImageUploader(props, ref) {
  const { id, onChange, ...otherProps } = props

  const [image, setImage] = useState<File>()
  const previewImage = useFilePreview(image)

  const renderIconAndImageTypes = () => {
    if (!!previewImage) return null

    return (
      <div
        data-testid='img-uploader-icon-and-image-types'
        className='flex flex-col items-center justify-center gap-2'
      >
        <DocumentPlusIcon className='h-10 w-10' />
        <p className='text-center text-xs'>
          PNG, JPG, JPEG
          <br />
        </p>
      </div>
    )
  }

  const renderPreviewImage = () => {
    if (!previewImage) return null

    return (
      <Image
        fill
        data-testid='img-uploader-preview-image'
        alt='Preview image'
        src={previewImage}
        style={{ objectFit: 'cover' }}
      />
    )
  }

  const customOnChange = (e: React.ChangeEvent<HTMLInputElement>) => {
    setImage(e.target.files?.[0])
    onChange?.(e)
  }

  return (
    <label
      htmlFor={id}
      className='relative flex aspect-square cursor-pointer flex-col items-center justify-center overflow-hidden rounded-md p-5 ring-1 ring-primary-500 hover:ring-2 hover:ring-primary-900 sm:aspect-auto sm:w-1/3'
    >
      {renderIconAndImageTypes()}
      {renderPreviewImage()}
      <input
        id={id}
        ref={ref}
        type='file'
        className='hidden'
        onChange={customOnChange}
        data-testid='img-uploader-input'
        accept='image/png, image/jpeg, image/jpg'
        {...otherProps}
      />
    </label>
  )
})
