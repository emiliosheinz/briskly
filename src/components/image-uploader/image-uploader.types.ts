export type ImageUploaderProps = {
  id: string
  error?: string
  defaultValue?: string
} & Omit<
  React.DetailedHTMLProps<
    React.InputHTMLAttributes<HTMLInputElement>,
    HTMLInputElement
  >,
  'type' | 'id' | 'defaultValue'
>
