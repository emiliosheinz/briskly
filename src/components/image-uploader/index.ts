export { ImageUploader } from './image-uploader.component'
