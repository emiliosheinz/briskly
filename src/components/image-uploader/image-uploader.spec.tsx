import { faker } from '@faker-js/faker'

import {
  fireEvent,
  render,
  screen,
  waitFor,
} from '~/utils/tests/react-testing-library'

import { ImageUploader } from './image-uploader.component'

jest.mock('~/hooks/use-file-preview', () => ({
  useFilePreview: (file?: File) => (file ? faker.internet.url() : ''),
}))

const makeFile = () =>
  new File(faker.random.words(10).split(' '), faker.random.word(), {
    type: 'image/png',
  })

describe('ImageUploader Component', () => {
  it('should be properly rendered', () => {
    const component = render(<ImageUploader id='id' />)

    expect(component.asFragment()).toMatchSnapshot()
  })

  it('should not show preview', () => {
    render(<ImageUploader id='id' />)

    const preview = screen.queryByTestId('img-uploader-preview-image')
    const iconAndImageTypes = screen.getByTestId(
      'img-uploader-icon-and-image-types',
    )

    expect(preview).not.toBeInTheDocument()
    expect(iconAndImageTypes).toBeInTheDocument()
  })

  it('should show preview when image is selected', async () => {
    const file = makeFile()

    render(<ImageUploader id='id' />)

    const imageInput = screen.getByTestId('img-uploader-input')

    await waitFor(() =>
      fireEvent.change(imageInput, {
        target: { files: [file] },
      }),
    )

    const preview = screen.getByTestId('img-uploader-preview-image')
    const iconAndImageTypes = screen.queryByTestId(
      'img-uploader-icon-and-image-types',
    )

    expect(preview).toBeInTheDocument()
    expect(iconAndImageTypes).not.toBeInTheDocument()
  })
})
