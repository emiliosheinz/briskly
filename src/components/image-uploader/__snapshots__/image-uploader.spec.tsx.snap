// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ImageUploader Component should be properly rendered 1`] = `
<DocumentFragment>
  <label
    class="relative flex aspect-square cursor-pointer flex-col items-center justify-center overflow-hidden rounded-md p-5 ring-1 ring-primary-500 hover:ring-2 hover:ring-primary-900 sm:aspect-auto sm:w-1/3"
    for="id"
  >
    <div
      class="flex flex-col items-center justify-center gap-2"
      data-testid="img-uploader-icon-and-image-types"
    >
      <svg
        aria-hidden="true"
        class="h-10 w-10"
        fill="none"
        stroke="currentColor"
        stroke-width="1.5"
        viewBox="0 0 24 24"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M19.5 14.25v-2.625a3.375 3.375 0 00-3.375-3.375h-1.5A1.125 1.125 0 0113.5 7.125v-1.5a3.375 3.375 0 00-3.375-3.375H8.25m3.75 9v6m3-3H9m1.5-12H5.625c-.621 0-1.125.504-1.125 1.125v17.25c0 .621.504 1.125 1.125 1.125h12.75c.621 0 1.125-.504 1.125-1.125V11.25a9 9 0 00-9-9z"
          stroke-linecap="round"
          stroke-linejoin="round"
        />
      </svg>
      <p
        class="text-center text-xs"
      >
        PNG, JPG, JPEG
        <br />
      </p>
    </div>
    <input
      accept="image/png, image/jpeg, image/jpg"
      class="hidden"
      data-testid="img-uploader-input"
      id="id"
      type="file"
    />
  </label>
</DocumentFragment>
`;
